### PR TITLE
Change password overhaul

### PR DESCRIPTION
**Pull request by Bettelstab**

### CODE REVIEW

**Comment by Bettelstab** [comment]
Not yet done: changed wording for legacy

File: src/request/change-password/ChangePassword.js [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+class ChangePassword {
+    /**
+     * @param {Parsed<KeyguardRequest.SimpleRequest>} request
+     * @param {ChangePassword.resolve} resolve
+     * @param {reject} reject
+     */
+    constructor(request, resolve, reject) {
+        this._resolve = resolve;
+        this._request = request;
+        this._reject = reject;
```
Comment (sisou) [inline src/request/change-password/ChangePassword.js]: Order of params is off. `_request` should be moved above ` _resolve`.

File: src/request/change-password/ChangePassword.js [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+
+        /** @type {Key} */
+        // eslint-disable-next-line no-unused-expressions
+        this._key;
```
Comment (sisou) [inline src/request/change-password/ChangePassword.js]: Could you make this `{Key?}` and assign it `null` instead of the eslint-disable?

File: src/request/change-password/ChangePassword.js [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+
+        // Pages
+        /** @type {HTMLFormElement} */
+        this.$enterPassword = (document.getElementById(ChangePassword.Pages.ENTER_PASSWORD));
+        /** @type {HTMLFormElement} */
+        this.$setPassword = (document.getElementById(ChangePassword.Pages.SET_PASSWORD));
```
Comment (sisou) [inline src/request/change-password/ChangePassword.js]: The `$enterPassword` and `$setPassword` page variables are only used within the constructor, thus they don't need to be assigned to the instance but can be local `const` only.

File: src/request/change-password/ChangePassword.js [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+        this._progressIndicator = new ProgressIndicator(
+            document.querySelector(`#${ChangePassword.Pages.SET_PASSWORD} .progress-indicator`),
+            4,
+            2,
+        );
+        // eslint-disable-next-line no-new
+        new ProgressIndicator(
+            document.querySelector(`#${ChangePassword.Pages.DOWNLOAD_FILE} .progress-indicator`),
+            4,
+            4,
```
Comment (sisou) [inline src/request/change-password/ChangePassword.js]: Please add a comment why this jumps from 2 to 4 and leaves out step 3 (because step 3 is used below for the password repetition).

File: src/request/change-password/ChangePassword.js [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+        this._passwordGetter.on(PassphraseBox.Events.SUBMIT, this._unlock.bind(this));
+
+        this._passwordSetter.on(PassphraseSetterBox.Events.ENTERED, () => {
+            let colorClass = '';
+            if (this._key.secret instanceof Nimiq.Entropy) {
+                const color = Iqons.getBackgroundColorIndex(
+                    new Nimiq.Address(
+                        // use color of first address as loginFile color
+                        this._key.deriveAddress(Constants.DEFAULT_DERIVATION_PATH).serialize(),
+                    ).toUserFriendlyAddress(),
```
Comment (sisou) [inline src/request/change-password/ChangePassword.js]: #229 adds `this._key.defaultAddress` to easily get the first address of a key. Reminder to replace this whole block with

```js
this._key.defaultAddress.toUserFriendlyAddress()
```

File: src/request/change-password/ChangePassword.js [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+            if (this._key.secret instanceof Nimiq.PrivateKey) {
+                this._resolve({ success: true });
+                return;
+            }
+
+            // Prepare Login File for download
+
+            const firstAddress = new Nimiq.Address(
+                this._key.deriveAddress(Constants.DEFAULT_DERIVATION_PATH).serialize(),
+            );
```
Comment (sisou) [inline src/request/change-password/ChangePassword.js]: Same here with the `key.defaultAddress`.

File: src/request/change-password/ChangePassword.js [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+
+            const encryptedEntropy = await this._key.secret.exportEncrypted(passwordBytes);
+
+            downloadLoginFile.setEncryptedEntropy(
+                /** @type {Nimiq.SerialBuffer} */ (encryptedEntropy),
```
Comment (sisou) [inline src/request/change-password/ChangePassword.js]: Maybe, instead of double-encryption, we could read the encrypted entropy from the database after it was stored and hand it to the LoginFile from there.

File: src/request/change-password/ChangePassword.js [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+            if (e.message === 'Invalid key') {
+                TopLevelApi.setLoading(false);
+                this._passwordGetter.onPassphraseIncorrect();
+                return;
+            }
+            this._reject(new Errors.CoreError(e));
+            return;
+        }
+        if (!key) {
+            this._reject(new Errors.KeyNotFoundError());
```
Comment (sisou) [inline src/request/change-password/ChangePassword.js]: _Technically_, key-existance is already checked during request parsing. I know that this check is also for Typescript, but it's still redundant and this code should never get triggered. I think we should make this error into something like a `ShouldNeverHappenError`, so that we are alerted when unexpected code paths are triggered. @nibhar?

File: src/request/change-password/ChangePassword.js [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+        TopLevelApi.setLoading(false);
+    }
+
+    /**
+     * @param {string} phrase
+     */
+    async _finish(phrase) {
+        TopLevelApi.setLoading(true);
+        if (!this._key) {
+            this._reject(new Errors.KeyguardError('Bypassed Password'));
```
Comment (sisou) [inline src/request/change-password/ChangePassword.js]:
```suggestion
            this._reject(new Errors.KeyguardError('Bypassed password setter'));
```
(Please don't use Githubs `Accept this suggestion` feature, but instead do this change yourself together with other changes in one commit.)

@nibhar This would also be a good place for the `ShouldNeverHappenError` ;)

File: src/request/change-password/ChangePassword.js [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+        if (TopLevelApi.getDocumentWidth() > Constants.MIN_WIDTH_FOR_AUTOFOCUS) {
+            this._passwordSetter.focus();
+        }
+        TopLevelApi.setLoading(false);
+    }
+
+    /**
+     * @param {string} phrase
+     */
+    async _finish(phrase) {
```
Comment (sisou) [inline src/request/change-password/ChangePassword.js]: Please rename `phrase` to `newPassword`.

File: src/request/change-password/ChangePassword.js [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+        window.location.hash = ChangePassword.Pages.ENTER_PASSWORD;
+        if (TopLevelApi.getDocumentWidth() > Constants.MIN_WIDTH_FOR_AUTOFOCUS) {
+            this._passwordGetter.focus();
+        }
+    }
+
+    /**
+     * @param {string} password
+     */
+    async _unlock(password) {
```
Comment (sisou) [inline src/request/change-password/ChangePassword.js]: The parameter should be named `oldPassword` instead, to differentiate between the passwords (for code clarity).

File: src/request/change-password/index.html [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+    <script defer src="../../components/LoginFileIcon.js"></script>
+    <script defer src="../../components/ProgressIndicator.js"></script>
+    <script defer src="../../components/PassphraseSetterBox.js"></script>
+    <script defer src="ChangePassword.js"></script>
+    <script defer src="ChangePasswordApi.js"></script>
+    <script defer src="index.js"></script>
+
+    <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Muli:400,600,700">
+    <link rel="stylesheet"
+        href="https://fonts.googleapis.com/css?family=Fira+Mono&text=0123456789ABCDEFGHJKLMNPQRSTUVXY">
```
Comment (sisou) [inline src/request/change-password/index.html]: There's really no need to have this in two lines, because no other index.html does it (the monospace font might even be unnecessary in this request? Is it even during change-password?)

File: src/request/change-password/index.html [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+
+<body class="loading">
+    <div class="header-top">
+        <div class="nimiq-app-name">
+            <span class="nimiq-logo"></span>
+        </div>
+    </div>
+
+    <div id="app">
+        <div class="flex-grow"></div>
```
Comment (sisou) [inline src/request/change-password/index.html]: The `flex-grow` elements should be unnecessary now with recent `common.css` fixes.

File: src/request/change-password/index.html [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+                <div class="progress-indicator"></div>
+                <a tabindex="0" class="page-header-back-button nq-icon arrow-left"></a>
+                <h1 class="nq-h1" data-i18n="change-password-set-password-heading">Create a new password</h1>
+            </div>
+
+            <div class="page-body nq-card-body">
+                <p data-i18n='change-password-set-password-text' class="nq-light-blue nq-text center">Secure your new Login File.</p>
+                <div class="flex-grow"></div>
+                <div class="nq-icon login-file-icon"></div>
+                <div class="flex-grow"></div>
```
Comment (sisou) [inline src/request/change-password/index.html]: The alignment can be managed in CSS without the need for `flex-grow` elements.

File: src/request/change-password/index.html [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+                <a tabindex="0" class="page-header-back-button nq-icon arrow-left"></a>
+                <h1 class="nq-h1" data-i18n="change-password-download-login-file">Download new Login File</h1>
+            </div>
+
+            <div class="page-body nq-card-body">
+                <a class="download-login-file"></a>
+            </div>
+        </div>
+
+        <div id="unlock-account" class="page nq-card">
```
Comment (sisou) [inline src/request/change-password/index.html]: I beleve this page is unused.

File: src/request/change-password/index.html [hunk 16]
```diff
@@ -0,0 +1,175 @@
...
+                <form class="password-box"></form>
+            </div>
+        </div>
+
+        <button class="global-close nq-button-s display-none">
+            <span class="nq-icon arrow-left"></span>
+            <span id="global-close-text"></span>
+        </button>
+
+        <div class="flex-grow"></div>
```
Comment (sisou) [inline src/request/change-password/index.html]: Same here: I think this is not necessary anymore.

File: src/translations/index.js [hunk 17]
```diff
@@ -244,14 +250,14 @@ const TRANSLATIONS = {
         'export-button-words': 'Wiederherstellungswörter anzeigen',
         'export-button-file': 'Schlüsseldatei herunterladen',
         'export-finish': 'Export abschließen',
-        'export-more-options-heading': 'Zusätzliche Export Optionen',
+        'export-more-options-heading': 'Zusätzliche Exportoptionen',
```
Comment (sisou) [inline src/translations/index.js]: Unrelated change.

File: src/components/PassphraseSetterBox.js [hunk 18]
```diff
@@ -61,6 +64,10 @@ class PassphraseSetterBox extends Nimiq.Observable {
         /** @type {HTMLButtonElement} */
         ($el.querySelector('button.submit')).classList.add('nq-button', 'inverse', options.bgColor);
 
+        if (options.hideSkip) {
+            ($el.removeChild(/** @type {HTMLAnchorElement} */($el.querySelector('.password-skip'))));
+        }
```
[the remark this answers is not in the document]
Comment (Bettelstab) [reply src/components/PassphraseSetterBox.js]: Sure. I did it like this to keep the htmlString in one piece. But I don't have a strong opinion about it.

File: src/request/change-password/ChangePassword.js (hunk 2 above)
Comment (Bettelstab) [reply src/request/change-password/ChangePassword.js]: But then I have to cast it later on, which adds much more noise.

Comment (sisou) [reply src/request/change-password/ChangePassword.js]: But `_key` _is_ `null` until the account is unlocked, so this is even _technically_ wrong.

Comment (Bettelstab) [reply src/request/change-password/ChangePassword.js]: Agreed. Found another solution.

File: src/request/change-password/ChangePassword.js (hunk 8 above)
Comment (Bettelstab) [reply src/request/change-password/ChangePassword.js]: Maybe it got lost in the meantime, which `shouldNeverHappen` of course.

File: src/request/change-password/index.html (hunk 12 above)
Comment (Bettelstab) [reply src/request/change-password/index.html]: Sorry, for some reason my VS Code thinks it's a good idea.
You're right, no address, no Fira. I'll remove it.

File: src/request/change-password/ChangePassword.js (hunk 7 above)
Comment (Bettelstab) [reply src/request/change-password/ChangePassword.js]: How do I do that?

File: src/request/change-password/ChangePassword.js [hunk 19]
```diff
@@ -0,0 +1,203 @@
...
+        }
+        this._loginFileIcon.lock(colorClass);
+        this._progressIndicator.setStep(3);
+    }
+
+    /**
+     * @param {string} newPassword
+     */
+    async _finish(newPassword) {
+        const passwordBytes = Utf8Tools.stringToUtf8ByteArray(newPassword);
```
Comment (sisou) [inline src/request/change-password/ChangePassword.js]: I think you accidentally lost the `TopLevelApi.setLoading(true)` call here in you last commit.

File: src/request/change-password/ChangePassword.js [hunk 20]
```diff
@@ -179,28 +136,63 @@ class ChangePassword {
...
-     * @param {string} phrase
+     * @param {string} newPassword
      */
-    async _finish(phrase) {
-        TopLevelApi.setLoading(true);
-        if (!this._key) {
-            this._reject(new Errors.KeyguardError('Bypassed Password'));
-            return;
-        }
+    async _finish(newPassword) {
```
Comment (sisou) [inline src/request/change-password/ChangePassword.js]: This function does not actually _finish_ the request for entropies. Those continue to the LoginFile download page. Thus `_finish` might be misleading.

File: src/request/change-password/ChangePassword.js (hunk 19 above)
Comment (Bettelstab) [reply src/request/change-password/ChangePassword.js]: Actually, it was missing before already. Finish was never called. But yeah, it's missing.

File: src/request/change-password/ChangePassword.js [hunk 21]
```diff
@@ -0,0 +1,212 @@
...
+            const colorString = LoginFile.CONFIG[color].name;
+            colorClass = `nq-${colorString}-bg`;
+        }
+        this._loginFileIcon.lock(colorClass);
+        this._progressIndicator.setStep(3);
+    }
+
+    /**
+     * Called after new password was entered second time.
+     * @param {string} newPassword
```
Comment (nibhar) [inline src/request/change-password/ChangePassword.js]: In case we want to allow for no password as new password this becomes 
```suggestion
     * @param {string} [newPassword]
```
That would also mean, that the test for 
```
if (this.key.secret instanceof Nimiq.PrivateKey) {
```
is no longer enough as unencrypted Entropies are also exempt from Login Files.

File: src/request/change-password/ChangePassword.js [hunk 22]
```diff
@@ -0,0 +1,212 @@
...
+        );
+
+        const encryptedEntropy = await this.key.secret.exportEncrypted(passwordBytes);
+
+        this._downloadLoginFile.setEncryptedEntropy(
+            /** @type {Nimiq.SerialBuffer} */(encryptedEntropy),
+            firstAddress,
+        );
+
+        this._downloadLoginFile.on(DownloadLoginFile.Events.DOWNLOADED, () => {
```
Comment (nibhar) [inline src/request/change-password/ChangePassword.js]: what happens if I go back and fourth between the login file and the password?

I think the event handling here should go to the constructor.

File: src/request/change-password/index.html [hunk 23]
```diff
@@ -62,53 +71,79 @@
...
+                    <svg class="nq-icon"><use xlink:href="../../../node_modules/@nimiq/style/nimiq-style.icons.svg#nq-arrow-left"/></svg>
+                </a>
+                <h1 data-i18n="change-password-heading" class="nq-h1">Confirm old password</h1>
             </div>
 
             <div class="page-body nq-card-body">
-                <p data-i18n="change-paragraph" class="nq-text">
-                    Before entering your new password, you must first unlock your account.
+                <p data-i18n="change-password-paragraph" class="nq-text nq-orange center">
+                    Changing the password will create a new Login File that replaces the current one.
```
Comment (nibhar) [inline src/request/change-password/index.html]: So is there no replacement text for this in case of a `Nimiq.PrivateKey`?

File: src/request/change-password/index.html [hunk 24]
```diff
@@ -62,53 +71,79 @@
...
                 </a>
-                <h1 data-i18n="change-passphrase-heading" class="nq-h1">Change your password</h1>
+                <h1 class="nq-h1" data-i18n="change-password-set-password-heading">Create a new password</h1>
             </div>
 
             <div class="page-body nq-card-body">
-                <p data-i18n="change-set-paragraph" class="nq-text">
-                    Now please enter your new password and repeat it for confirmation.
-                </p>
+                <p data-i18n='change-password-set-password-text' class="nq-light-blue nq-text center">Secure your new Login File.</p>
```
Comment (nibhar) [inline src/request/change-password/index.html]: Also no replacement text?
